### PR TITLE
[fastdds] Restructure Fast-DDS folder and prepare 2.3.4

### PR DESCRIPTION
**Pull request by dotChris90**
Fast-DDS package / 2.3.4

want to add the 2.3.4 package of Fast-DDS and also reference the foonathan-memory package with options so the examples are executable --> see https://github.com/conan-io/conan-center-index/issues/6731. 

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/fast-dds/all/conanfile.py [hunk 1]
```diff
@@ -117,6 +117,7 @@ def requirements(self):
         self.requires("asio/1.18.2")
         self.requires("fast-cdr/1.0.21")
         self.requires("foonathan-memory/0.7.0")
+        self.options["foonathan-memory"].with_sizecheck = False
```
Comment (jgsogo) [inline recipes/fast-dds/all/conanfile.py:120]: Hi! In ConanCenter we are generating only the packages with the default option values (except `shared` and `fPIC`, we generate packages for `True` and `False` values). If you modify these defaults, the CI will fail with a _missing package_ exception.

Given that this requirement is required for tests (and we are not building the tests in CCI), I would suggest the following: add option `test/with_tests/build_test` with default value `False`. Conditioned on that option, you can add this requirement and build the tests (better if we can avoid building them by default, even if it requires a patch).

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: thanks for pointing it out. I was yesterday really wondering ^^' but it makes absolute sence now.

Comment (smasihemami) [reply recipes/fast-dds/all/conanfile.py:120]: On trying to make a local version of 2.3.4 by "conan create" command, I'm getting this error:

ERROR: foonathan-memory/0.7.0: option 'with_sizecheck' doesn't exist
Possible options are ['shared', 'with_tools']

Am I missing something here?

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: I will check it today too - i did PR on weekend for this foonathan_memory change (adding new option with_sizecheck). 
Fast-DDS 2.3.4 was my local cache for creation of package - so i will validate it on my local pc today too.

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: that's strange - you are right. "conan inspect foonathan-memory/0.7.0@_/_ -r conan-center" 
gives 

name: foonathan-memory
version: 0.7.0
url: https://github.com/conan-io/conan-center-index
homepage: https://github.com/foonathan/memory
license: Zlib
author: None
description: STL compatible C++ memory allocator library
topics: ('conan', 'memory', 'STL', 'RawAllocator')
generators: cmake
exports: None
exports_sources: ['patches/**', 'CMakeLists.txt']
short_paths: False
apply_env: True
build_policy: None
revision_mode: hash
settings: ('os', 'compiler', 'build_type', 'arch')
options:
    fPIC: [True, False]
    shared: [True, False]
    with_tools: [True, False]
default_options:
    fPIC: True
    shared: False
    with_tools: False
deprecated: None

@smasihemami is there the possbility in CC Artifactory is old package present?

Comment (smasihemami) [reply recipes/fast-dds/all/conanfile.py:120]: Not really sure. A package index should be immutable in principle, meaning a duplicate should not be allowed. Not sure how CC Artifactory works behind the scenes, and what would be the expected behaviour if one pushes a package twice with the same version.
Did foonathan-memory/0.7.0 exist before your pull request on CC Artifactory?

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: At weekend the Fast-DDS branch and the Foonathan branch exist on my conan-center github fork. 

I am pretty sure that i did the Fast-DDS PR after i got information "hey PR - foonathan" merged. 

Is there the possbility i can trigger the jenkins servers again for the foonathan memory or is this just possible with a "fake -PR"?

Comment (jgsogo) [reply recipes/fast-dds/all/conanfile.py:120]: Are you using the proper remote (`conan remote list`)? We have stop uploading packages to `conan.bintray.com`, new ones are only uploaded to `center.conan.io`. You can read about it: https://blog.conan.io/2021/06/10/New-conan-release-1-37.html

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: @jgsogo,  @smasihemami ..... my bad ..... sorry =.= .... the PC i used to quick check referencing the old one .... jgsogo is totally right - the new reference also referencing the new package with the option.

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: @jgsogo sorry to ask stupid questions again :D - about the jenkins default packages that are build and set to artifactory. I made an observation and would like to ask if i saw correctly. 

Maybe background : 
- foonathan_memory is used by fast-dds and fast-dds requires a non default option (that's why with_sizecheck = False)
- i decided to set "with_sizecheck = True" as default option because the underlying cmake option also is True as default (not sure if it's a CCI rule but i saw that most conan packages set their default option corresponding to cmake options) 
- that's why currently we have the default mismatch 

Now local PC i did in the in the conanfile.py of test_package folder of foonathan_memory package  : 
self.options["foonathan-memory"].with_sizecheck = False

and i saw that in local cache ~/.conan/data/ just one hash-value is generated for foonathan_package 

then i set : 
self.options["foonathan-memory"].with_sizecheck = True

and i saw in local cache ~/.conan/data that a 2nd hash value is present (i guess it represents build with option false and true). 

Is this a hack or can i truly "effect the Conan Center packages" which are in CC Artifactory by this? 
If so maybe i simple can set the test_package conanfile with option "with_sizecheck = False" - then this package become part of conan center artifactory and fast-dds would not throw missing package.

Comment (smasihemami) [reply recipes/fast-dds/all/conanfile.py:120]: @dotChris90  Just noticed another thing. According to this page (Under Manual Installation): https://github.com/eProsima/Fast-DDS/tree/2.3.3 , They (eProsima) are not using https://github.com/foonathan/memory (which is mentioned in the recipe for foonathan-memory/0.7.0) but using https://github.com/eProsima/foonathan_memory_vendor.git
Don't really know what the difference is. And the one they're using has the latest version of 0.5.0

Yet another thing which is not totally clear for me is what git revision is used for creation of foonathan-memory@0.7.0 on CC Artifactory

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: @smasihemami that is correct so far. and actually foonathan_memory here is bigger thing ..... unfortunately =.= 

If you looking to their repo foonathan_memory_vendor you will see that they mostly just setting the build option for foonathan_memory, do some toolchain fixes (affecting cross compilation for android) and then referencing the foonathan_memory repo at specific commit --> https://github.com/eProsima/foonathan_memory_vendor/blob/0be8f746ee33b3366528d2533087754ce14732b0/CMakeLists.txt#L79 

This commit is bettwen releases 0.6.2 and 0.7.0 .... and in my honest opinion a package should reference releases and not random commits .... they explained on their repo already to me that they planning to migrate to 0.7.0 but not done yet. 

Long story short - i looked into their patch https://github.com/eProsima/foonathan_memory_vendor/blob/master/CMakeLists.txt.patch and their repo in general and didn't find any reason to not directly reference the true repository - especially to avoid packages that contains no source code but just meta information ....

Comment (jgsogo) [reply recipes/fast-dds/all/conanfile.py:120]: Hi, @dotChris90. No, you won't be able to hack CCI that way (I hope so 😄 ). In CCI we are running `conan install <your/conanfile.py> --build=fast-dds` so it will only trigger a build for `fast-dds`, the rest of requirements will be reported as missing. On top of that, we are running several checks to ensure that the packageID generated is the one we are expecting and we will only upload that packageID.

> i decided to set "with_sizecheck = True" as default option because the underlying cmake option also is True as default (not sure if it's a CCI rule but i saw that most conan packages set their default option corresponding to cmake options)

Yes, this is usually the rule of thumb, we try to initialize the options with the same defaults the library is using. Nevertheless it is perfectly ok to change them if it makes sense in the context of CCI. In the end, users can always generate the package from sources with very little effort.

Here, if it is absolutely needed to generate `foonathan-memory` with a different option value, we need to think something, because currently CCI doesn't implement any workaround (or feature) to handle that situation.

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: @jgsogo maybe i would suggest the following - i do a PR again with foonathan_memory and change the default option to False so fast-dds build can go on.

i currently saw in https://conan.io/center/foonathan-memory?tab=usedby that foonathan_memory is referenced just by fast-dds. 

In my honest opinion i think Fast-DDS is an industrial cool project since it's the most complete DDS implementation i ever saw. So i think it's ok if we here breaking the "rule of thumb" for foonathan_memory and simple go the easy path : change the default value of option to False even cmake option is true. 

i will keep this PR open and after foonathan_memory PR is mergt trigger it again. 

just a question of curiosity - was there never the need for CCI package to have "conan install --build=missing .."? :D as i said just curious.

Comment (jgsogo) [reply recipes/fast-dds/all/conanfile.py:120]: Open the PR with the problematic and community will help. Maybe they can suggest a different approach.

> just a question of curiosity - was there never the need for CCI package to have "conan install --build=missing .."? :D as i said just curious.

Up to now, we are not running `--build=missing` so it hasn't been needed for all the packages that are already in ConanCenter (maybe there is some pending PR with this same problem). Sometimes, we have faced similar situations, especially for options that were activating some feature that was only available in some operating system, but someone came up with a really nice workaround: if the feature is not available, then remove the option and use `self.options.get_safe('<option>', some-default)`, that way we have _two different values_ for the same option depending on the OS. I'm afraid this is not the case here.

Let's see what happens in that PR.

Comment (prince-chrismc) [reply recipes/fast-dds/all/conanfile.py:120]: > Up to now, we are not running `--build=missing` so it hasn't been needed for all the packages that are already in ConanCenter

I think we just worked around it thus far... and disabled feature of downstream projects to avoid those "not set by default options" 

Yet another feature request for CCI but having `--build=missing` and an `upload --all` would be great to enable more options (though they would only be tested by downstream recipes)

Comment (prince-chrismc) [reply recipes/fast-dds/all/conanfile.py:120]: This now needs to be moved to validate incase it changes in the future

Comment (dotChris90) [reply recipes/fast-dds/all/conanfile.py:120]: i will do the change - yes validate makes sense here.

Comment (danimtb) [reply recipes/fast-dds/all/conanfile.py:120]: can we resolve this conversation then?

**Review by SSE4 — CHANGES_REQUESTED** [review]
please remove:
```
self.options["foonathan-memory"].with_sizecheck = False
```
and add a check to the `validate` method as suggested

**Comment by jgsogo** [comment]
@dotChris90 , I think we just need one minor change here to be ready...

**Comment by prince-chrismc** [comment]
You'll need to retrigger CI, close the pr wait 10s and then re-open it 🔁

**Review by prince-chrismc — APPROVED** [review]
The test package is larger than I'd like but it seems fairly minimal given what it needs to do

**Comment by dotChris90** [comment]
I create a new PR - first restructure folder then add 2.3.4
